ngraph/core/reference/include/ngraph/runtime/reference/fake_quantize.hpp

```diff
@@ -53,182 +55,243 @@ namespace ngraph
+                        case Bound::trivial: val = *bounds; break;
+                        case Bound::aligned: val = bounds[idx]; break;
+                        case Bound::broadcast:
                         {
-                            NGRAPH_CHECK(idx >= index_offset, "Incorrect index offset value!");
+                            const size_t index_offset =
+                                calc_full_broadcast_offset(current_dim, row_strides);
+                            NGRAPH_CHECK(0 <= index_offset && index_offset <= idx,
+                                         "Incorrect index offset value!");
+                            val = bounds[idx - index_offset];
+                            break;
+                        }
                         }
-                        val = data[idx - index_offset];
+                        return val;
                     }
-                    return val;
+
+                private:
+                    Bound bound;
+                    std::vector<size_t> row_strides;
+                    const T* const bounds;
                 };
-                for (size_t i = 0; i < shape_size(arg_shape); ++i)
+
+                template <typename T>
+                inline T quantize(const T& arg,
+                                  const T& in_low,
+                                  const T& in_high,
+                                  const T& out_low,
+                                  const T& out_high,
+                                  const size_t& levels)
                 {
-                    T in_low_val = get_value(
-                        in_low_trivial_broadcast, in_low_aligned, in_low, i, in_low_offsets);
-                    T in_high_val = get_value(
-                        in_high_trivial_broadcast, in_high_aligned, in_high, i, in_high_offsets);
-                    T out_low_val = get_value(
-                        out_low_trivial_broadcast, out_low_aligned, out_low, i, out_low_offsets);
-                    T out_high_val = get_value(out_high_trivial_broadcast,
-                                               out_high_aligned,
-                                               out_high,
-                                               i,
-                                               out_high_offsets);
-                    if (arg[i] <= std::min(in_low_val, in_high_val))
+                    if (arg <= std::min(in_low, in_high))
+                    {
+                        return out_low;
+                    }
+                    else if (arg > std::max(in_low, in_high))
                     {
-                        out[i] = out_low_val;
+                        return out_high;
                     }
-                    else if (arg[i] > std::max(in_low_val, in_high_val))
+                    return std::nearbyint((arg - in_low) / (in_high - in_low) * (levels - 1)) /
+                               (levels - 1) * (out_high - out_low) +
+                           out_low;
```

Comment:
Are we checking somewhere that `(out_high - out_low)` is not `0`?

Reply:
There is no check if `(in_high - in_low)` or `(out_high - out_low)` is zero (division by zero). 
Check should be add or is this UB?

Reply:
Oh I see, there is probably no need as only `(levels-1)` not `((levels - 1) * (out_high - out_low))` is the divisor here.
Also it looks like the `min` and `max` above handle `(in_high - in_low)` zero case.